executor/aggregate.go

```diff
@@ -45,39 +106,490 @@ type HashAggExec struct {
+		for i, af := range w.aggFuncs {
+			if af.Update(aggEvalCtxs[i], sc, row) != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
 	return nil
 }
 
+// shuffleIntermData shuffles the intermediate data of partial workers to corresponded final workers.
+// We only support parallel execution for single-machine, so process of encode and decode can be skipped.
+func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, finalConcurrency int) {
+	groupKeysSlice := make([][][]byte, finalConcurrency)
+	for groupKey := range w.aggCtxsMap {
+		groupKeyBytes := []byte(groupKey)
+		finalWorkerIdx := int(murmur3.Sum32(groupKeyBytes)) % finalConcurrency
+		if groupKeysSlice[finalWorkerIdx] == nil {
+			groupKeysSlice[finalWorkerIdx] = make([][]byte, 0, len(w.aggCtxsMap)/finalConcurrency)
+		}
+		groupKeysSlice[finalWorkerIdx] = append(groupKeysSlice[finalWorkerIdx], groupKeyBytes)
+	}
+
+	for i := range groupKeysSlice {
+		if groupKeysSlice[i] == nil {
+			continue
+		}
+		w.outputChs[i] <- &HashAggIntermData{
+			groupKeys:   groupKeysSlice[i],
+			groupCtxMap: w.aggCtxsMap,
+		}
+	}
+}
+
+// getGroupKey evaluates the group items and args of aggregate functions.
+func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, row chunk.Row) ([]byte, error) {
+	w.groupValDatums = w.groupValDatums[:0]
+	for _, item := range w.groupByItems {
+		v, err := item.Eval(row)
+		if item.GetType().Tp == mysql.TypeNewDecimal {
+			v.SetLength(0)
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		w.groupValDatums = append(w.groupValDatums, v)
+	}
+	var err error
+	w.groupKey, err = codec.EncodeValue(sc, w.groupKey[:0], w.groupValDatums...)
+	return w.groupKey, errors.Trace(err)
+}
+
+func (w baseHashAggWorker) getContext(sc *stmtctx.StatementContext, groupKey []byte, mapper aggCtxsMapper) []*aggregation.AggEvaluateContext {
+	aggCtxs, ok := mapper[string(groupKey)]
+	if !ok {
+		aggCtxs = make([]*aggregation.AggEvaluateContext, 0, len(w.aggFuncs))
+		for _, af := range w.aggFuncs {
+			aggCtxs = append(aggCtxs, af.CreateContext(sc))
+		}
+		mapper[string(groupKey)] = aggCtxs
+	}
+	return aggCtxs
+}
+
+func (w *HashAggFinalWorker) getPartialInput() (input *HashAggIntermData, ok bool) {
+	select {
+	case <-w.finishCh:
+		return nil, false
+	case input, ok = <-w.inputCh:
+		if !ok {
+			return nil, false
+		}
+	}
+	return
+}
+
+func (w *HashAggFinalWorker) consumeIntermData(sc *stmtctx.StatementContext) (err error) {
+	var (
+		input *HashAggIntermData
+		ok    bool
+	)
+	for {
+		if input, ok = w.getPartialInput(); !ok {
+			return nil
+		}
+		if w.intermDataRowsBuffer == nil {
+			w.intermDataRowsBuffer = make([]types.DatumRow, 0, w.maxChunkSize)
+		}
+		// Consume input in batches, size of every batch is less than w.maxChunkSize.
+		for reachEnd := false; !reachEnd; {
+			w.intermDataRowsBuffer, reachEnd = input.ToRows(sc, w.intermDataRowsBuffer, w.aggFuncs, w.maxChunkSize)
```

Comment:
`input.ToRows(sc, w.intermDataRowsBuffer[:0], w.aggFuncs, w.maxChunkSize)`, thus line 433 can be removed